### PR TITLE
Makefile: Verify golangci-lint integrity before using it

### LINKED ISSUE #64: Makefile should verify downloaded golangci-lint executables
### Description
Right now the makefile blindly downloads and executes the golangci-lint installation script and accepts whatever it downloads as the linter to use. We should store the signatures of both files in the makefile and verify that they match.

### Motivation
An attacker could compromise either file so that whoever builds the SDK runs arbitrary code.

### Exemplification
N/A

### Benefits
One security vulnerability down, _N_ to go.

### Possible Drawbacks
None.

### PR DESCRIPTION
**Pull request by magnusbaeck**
### Applicable Issues
Fixes #64

### Description of the Change
The makefile now verifies the SHA-256 checksums of the installation script for golangci-lint and the binary we're expecting to see in GOBIN afterwards. This makes it a bit more tedious to upgrade golangci-lint but reduces surprises.

### Alternate Designs
I considered splitting the single makefile rule into multiple rules,
```makefile
golangci-lint: installation-script
        installation-script > $@.unverified
        sha256sum - < $@.unverified
        mv $@.unverified $@

installation-script: installation-script.unverified
        sha256sum - < $?
        mv $@.unverified $@

installation-script.unverified:
        curl -o $@ ...
```
but figured that it was less readable for most people and didn't bring any major benefits.

### Benefits
If the golangci-lint installation script or the binaries downloaded by the script are compromised we'll detect it before it causes any damage.

### Possible Drawbacks
None.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I    have the right to submit it under the open source license    indicated in the file; or

(b) The contribution is based upon previous work that, to the best    of my knowledge, is covered under an appropriate open source    license and I have the right under that license to submit that    work with modifications, whether created in whole or in part    by me, under the same open source license (unless I am    permitted to submit under a different license), as indicated    in the file; or

(c) The contribution was provided directly to me by some other    person who certified (a), (b) or (c) and I have not modified    it.

(d) I understand and agree that this project and the contribution    are public and that a record of the contribution (including all    personal information I submit with it, including my sign-off) is    maintained indefinitely and may be redistributed consistent with    this project or the open source license(s) involved.

Signed-off-by: Magnus Bäck \<magnus.back@axis.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,7 +63,15 @@ clean:
 	$(GOCLEAN)
 	rm -f $(PROGRAM)
 
+# Download the installation script for golangci-lint, verify its SHA-256 digest,
+# run it if everything checks out, and verify the resulting binary.
 $(GOLANGCI_LINT):
-	curl -sfL \
-			https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh \
-		| sh -s -- -b $(GOBIN) $(GOLANGCI_LINT_VERSION)
+	mkdir -p $(dir $@)
```
Comment (JonasAlfredsson) [inline Makefile:69]: I am trying to wrap my head around what this line does.
`$@`: the name of the target being generated
So `$@` in this case is `$(GOLANGCI_LINT)`, which expands to `$(CURDIR)/bin/golangci-lint`
And I get the following when I just try the `dir` command:
```
dir $PWD/bin/golangci-lint
dir: cannot access '/home/jonasal/bin/golangci-lint': No such file or directory
```
while I get all the filenames and folders present in the case the path actually exist.

Is it something I am missing with this or will this fail in the ci case when the `bin/` folder has not been created?

File: Makefile [hunk 2]
```diff
@@ -61,7 +63,15 @@ clean:
...
+	curl -sSfL \
+		https://raw.githubusercontent.com/golangci/golangci-lint/$(GOLANGCI_LINT_VERSION)/install.sh \
+		> $@.install-script-unverified
+	echo "$(GOLANGCI_LINT_INSTALLATION_SHA256) $@.install-script-unverified" | sha256sum -c --quiet -
+	sh -s -- -b $(dir $@) $(GOLANGCI_LINT_VERSION) < $@.install-script-unverified
+	rm -f $@.install-script-unverified
```
Comment (JonasAlfredsson) [inline Makefile:75]: Should you append this line onto 73 like you do on 76?

File: Makefile (hunk 1 above)
Comment (magnusbaeck) [reply Makefile:69]: `$(dir ...)` isn't a call to a shell command but a call to a [make function](https://www.gnu.org/software/make/manual/html_node/File-Name-Functions.html#index-dir) that's basically equivalent to dirname(1). IOW, line 69 will expand to `mkdir -p $(CURDIR)/bin/`. If you want your command to actually contain a dollar sign for e.g. a shell variable reference you need `$$`. Have I ever forgotten this and been confused over my makefile not working to my satisfaction? Yes. Yes, I have.

Comment (JonasAlfredsson) [reply Makefile:69]: Alrighty! :)

File: Makefile (hunk 2 above)
Comment (magnusbaeck) [reply Makefile:75]: I pushed a commit for that just now but realized that we'd have to add it to line 74 as well if we don't want to risk leaving an unverified but non-executable file with "unverified" in the name. It gets pretty noisy and I'm not sure I think it's worth it.

This case is different from line 76, where a failure would leave us with an executable `$@` in place, i.e. a subsequent `make` invocation wouldn't cause this rule to run again and the bad binary would be run. That's dangerous, but for the installation script you have to work pretty hard to run an unverified one.

Comment (JonasAlfredsson) [reply Makefile:75]: Yeah, got a notification that Magnus pushed 0 commits! 
You know you are productive when you get an email saying that nothing has been done :P
But alright, I support your reasoning :)

**Review by JonasAlfredsson — APPROVED** [review]

**Review by felixhall — APPROVED** [review]

**Pull request merged**